### PR TITLE
Remove usage of grpc-gateway in otlpreceiver.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Http protocols are different than grpc

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: receiver/otlpreceiver/otlp.go [hunk 1]
```diff
@@ -191,40 +187,160 @@ func (r *otlpReceiver) Shutdown(ctx context.Context) error {
...
+		}).Methods(http.MethodPost).Headers("Content-Type", pbContentType)
+		// For backwards compatibility see https://github.com/open-telemetry/opentelemetry-collector/issues/1968
+		r.httpMux.HandleFunc("/v1/trace", func(resp http.ResponseWriter, req *http.Request) {
+			handleHttpTraces(resp, req, pbContentType, r.traceReceiver, tracesPbUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", pbContentType)
+		r.httpMux.HandleFunc("/v1/traces", func(resp http.ResponseWriter, req *http.Request) {
+			handleHttpTraces(resp, req, jsonContentType, r.traceReceiver, tracesJSONUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", jsonContentType)
+		// For backwards compatibility see https://github.com/open-telemetry/opentelemetry-collector/issues/1968
+		r.httpMux.HandleFunc("/v1/trace", func(resp http.ResponseWriter, req *http.Request) {
```
Comment (carlosalberto) [inline receiver/otlpreceiver/otlp.go:210]: +1

**Review by carlosalberto — APPROVED** [review]

**Review by jrcamp — APPROVED** [review]

File: receiver/otlpreceiver/otlphttp.go [hunk 2]
```diff
@@ -38,39 +38,59 @@ func (*xProtobufMarshaler) ContentType() string {
...
 	if contentType == "application/json" {
 		buf := new(bytes.Buffer)
-		err = jsonMarshaler.Marshal(buf, s.Proto())
-		msg = buf.Bytes()
+		err = jsonMarshaler.Marshal(buf, rsp)
+		// Nothing we can do with the error if we cannot write to the response.
+		_, _ = w.Write(buf.Bytes())
 	} else {
-		msg, err = proto.Marshal(s.Proto())
+		msg, err = proto.Marshal(rsp)
```
Comment (rakyll) [inline receiver/otlpreceiver/otlphttp.go:86]: Shouldn't we log these errors?

File: receiver/otlpreceiver/otlphttp.go [hunk 3]
```diff
@@ -38,39 +38,59 @@ func (*xProtobufMarshaler) ContentType() string {
...
 		msg = fallbackMsg
 		contentType = fallbackContentType
 		statusCode = http.StatusInternalServerError
 	}
-
-	w.Header().Set("Content-Type", contentType)
-	w.WriteHeader(statusCode)
-	w.Write(msg) // nolint:errcheck
+	// Nothing we can do with the error if we cannot write to the response.
+	_, _ = w.Write(msg)
```
Comment (rakyll) [inline receiver/otlpreceiver/otlphttp.go:95]: Ditto.

File: receiver/otlpreceiver/otlphttp.go [hunk 4]
```diff
@@ -38,39 +38,59 @@ func (*xProtobufMarshaler) ContentType() string {
...
+}
+
+// errorHandler encodes the HTTP error message inside a rpc.Status message as required
+// by the OTLP protocol.
+func errorHandler(w http.ResponseWriter, r *http.Request, errMsg string, statusCode int) {
+	writeErrorMsg(w, r.Header.Get("Content-Type"), errMsg, statusCode)
+}
+
+// Pre-computed status with code=Internal to be used in case of a marshaling error.
+var fallbackMsg = []byte(`{"code": 13, "message": "failed to marshal error message"}`)
```
Comment (rakyll) [inline receiver/otlpreceiver/otlphttp.go:71]: Give a more self representative name such as marshalingErrorMsg = = []byte(`{"code": 13, "message": "failed to marshal error message"}`)

File: receiver/otlpreceiver/otlp.go [hunk 5]
```diff
@@ -191,40 +187,160 @@ func (r *otlpReceiver) Shutdown(ctx context.Context) error {
...
+	rsp, err := metricsReceiver.Export(req.Context(), internal.MetricsToOtlp(md.InternalRep()))
+	if err != nil {
+		writeError(resp, contentType, err, http.StatusInternalServerError)
+		return
+	}
+
+	writeResponse(resp, contentType, http.StatusOK, rsp)
+}
+
+func handleHttpLogs(
```
Comment (rakyll) [inline receiver/otlpreceiver/otlp.go:317]: Ditto.

File: receiver/otlpreceiver/otlp.go [hunk 6]
```diff
@@ -191,40 +187,160 @@ func (r *otlpReceiver) Shutdown(ctx context.Context) error {
...
+			handleHttpLogs(w, req, pbContentType, r.logReceiver, logsPbUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", pbContentType)
+		r.httpMux.HandleFunc("/v1/logs", func(w http.ResponseWriter, req *http.Request) {
+			handleHttpLogs(w, req, jsonContentType, r.logReceiver, logsJSONUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", jsonContentType)
 	}
 	return nil
 }
+
+func handleHttpTraces(
```
Comment (rakyll) [inline receiver/otlpreceiver/otlp.go:255]: handleHTTPTraces

Btw, are there any non-HTTP handlers? We can simply call this handleTraces.

File: receiver/otlpreceiver/otlp.go [hunk 7]
```diff
@@ -191,40 +187,160 @@ func (r *otlpReceiver) Shutdown(ctx context.Context) error {
...
+	rsp, err := tracesReceiver.Export(req.Context(), internal.TracesToOtlp(td.InternalRep()))
+	if err != nil {
+		writeError(resp, contentType, err, http.StatusInternalServerError)
+		return
+	}
+
+	writeResponse(resp, contentType, http.StatusOK, rsp)
+}
+
+func handleHttpMetrics(
```
Comment (rakyll) [inline receiver/otlpreceiver/otlp.go:286]: Ditto

**Review by rakyll — APPROVED** [review]

File: receiver/otlpreceiver/otlp.go (hunk 6 above)
Comment (bogdandrutu) [reply receiver/otlpreceiver/otlp.go:255]: grpc

File: receiver/otlpreceiver/otlp.go (hunk 5 above)
Comment (bogdandrutu) [reply receiver/otlpreceiver/otlp.go:317]: Done.

File: receiver/otlpreceiver/otlp.go (hunk 7 above)
Comment (bogdandrutu) [reply receiver/otlpreceiver/otlp.go:286]: Done.

**Pull request merged**